quay/quay-operator
# PROJQUAY-880: Add Monitoring component to Quay Operator

**Issue:** https://issues.redhat.com/browse/PROJQUAY-880


**Changelog:** 

* Added a new `monitoring` component to the Quay Spec
* Added `Service` and `ServiceMonitor` for exporting Quay metrics to the in-cluster Prometheus
* Added  Grafana dashboard with default Graphs for viewing metrics realtime
* Added an Alert which triggers when there are too many crashes of the Quay pod
* Updated `ComponentKind` to use an enum instead of string
* Updated README docs for development 

**Testing:** 

1. Setup a new `QuayRegistry` 

```
apiVersion: quay.redhat.com/v1
kind: QuayRegistry
metadata:
  name: syed-test-quay
  namespace: syed-test-quay
spec:
  components:
    - kind: monitoring
      managed: true
    - kind: clair
      managed: false
```

2. Make sure that you can see `Quay` under the dashboards section on the Openshift Console.

<img width="1506" alt="Screen Shot 2021-03-18 at 9 40 38 AM" src="https://user-images.githubusercontent.com/55854/111635598-0ca3b480-87ce-11eb-8e63-bf9600472b46.png">

3. Create an `Org` in your Quay instance and some traffic to generate metrics. The following script should be able to generate enough traffic to display metrics on the dashboard

```
#!/bin/bash

set -x

ORG=$1

export OCP_RELEASE="4.7.2-x86_64"
export LOCAL_REGISTRY='<your-quay-instance>:443'
export LOCAL_REPOSITORY='$ORG/openshift4'
export PRODUCT_REPO='openshift-release-dev'
export LOCAL_SECRET_JSON="$PWD/pull-secrets-fixed.json"
export RELEASE_NAME="ocp-release"

oc adm -a ${LOCAL_SECRET_JSON} release mirror \
     --from=quay.io/${PRODUCT_REPO}/${RELEASE_NAME}:${OCP_RELEASE} \
     --to=${LOCAL_REGISTRY}/${LOCAL_REPOSITORY} \
     --to-release-image=${LOCAL_REGISTRY}/${LOCAL_REPOSITORY}:${OCP_RELEASE}
```

You need to get the `pull-secrets` JSON file from [OpenShift Container Platform downloads](https://access.redhat.com/downloads/content/290/) page and add the credentials to your Quay instance in that file. More information can be found [Here](https://docs.openshift.com/container-platform/4.2/installing/install_config/installing-restricted-networks-preparations.html)

Update the variables to reflect your setup and save the script as `mirror.sh` and run it using the following command:

```
./mirror.sh <your-org-name>
```

Once you start running the above script. You should see some data in the dashboard

![Screen Shot 2021-03-13 at 8 27 39 AM](https://user-images.githubusercontent.com/55854/111634592-0c56e980-87cd-11eb-8b74-6b07bf2b4aa2.png)

![Screen Shot 2021-03-13 at 8 30 31 AM](https://user-images.githubusercontent.com/55854/111634621-124cca80-87cd-11eb-8152-f9eca845b80e.png)
